plugins/modules/edpm_derive_pci_passthrough_devicespec.py

```diff
@@ -0,0 +1,650 @@
+    :return user_config_copy: Any non-nic partinioned cfg will be returned in this list
+    :return derived_config: All nic partinioned cfg will be returned after derivation in this list
+    """
+
+    user_config_copy = []
+    derived_config = []
+
+    allocated_pci = get_allocated_pci_addresses(system_configs)
+    nic_partition_pfs = get_sriov_nic_partition_pfs(system_configs)
+
+    for user_config in user_configs:
+        if 'address' in user_config and 'devname' in user_config:
+            msg = f"Both devname and address can't be present in {_PASSTHROUGH_DEVSPEC_KEY}"
+            raise InvalidConfigException(msg)
+
+        keys = ['address', 'product_id', 'devname']
+        if not any(k in user_config for k in keys):
+            # address or product_id or devname not present in user_config
+            pf = get_pf_name_from_phy_network(user_config['physical_network'], phys_dev_map)
+            user_config['address'] = get_pci_addr_from_ifname(pf)
+
+        if 'devname' in user_config:
+            if user_config['devname'] in nic_partition_pfs:
+                user_config['address'] = get_pci_addr_from_ifname(user_config['devname'])
+                del user_config['devname']
+            else:
+                user_config_copy.append(user_config)
+                continue
+        if ('address' in user_config
+                or 'product_id' in user_config):
+            passthrough_tmp, del_user_config = get_passthrough_config_for_all_pf(
+                user_config, system_configs, allocated_pci)
+
+            """ If del_user_config is set, do not add to derived_config or
+                user_config_copy
+            """
+            if not del_user_config:
+                if passthrough_tmp:
+                    derived_config.extend(passthrough_tmp)
+                else:
+                    user_config_copy.append(user_config)
+        else:
+            user_config_copy.append(user_config)
+    return (user_config_copy, derived_config)
+
+
+def get_nova_userconfigs(nova_conf_file):
+    configs = []
+    src_config = ""
+    # Currently only "device_spec" is valid
+    valid_spec = ('device_spec')
+
+    # Read user config from edpm_nova_config conf file
+    try:
+        with open(nova_conf_file) as nova_config:
+            for line in nova_config:
+                pf = line.strip()
+                pf = line.replace('\'', '\"')
+                if pf.startswith(valid_spec):
+                    # make a copy for replace after derive
+                    src_config += line
+                    obj = pf.split('=')[1]
+                    spec = json.loads(obj)
+                    configs.append(spec)
+    except IOError:
+        configs = []
+    return (configs, src_config)
+
+
+def _run_derive_pci(sriov_device_spec,
+                    sriov_nova_conf_file,
+                    sriov_phydev_map=None):
+    pci_passthrough = {}
+    system_configs = get_sriov_configs()
+    user_configs = sriov_device_spec
```

Comment:
@vcandapp remove this line and specify user_configs as _run_derive_pci function argument sriov_device_spec,?
since no user_configs updates done. only for reading purpose.

Reply:
Done